### DISCUSSION TITLE
Possible Mullvad-cli package for openwrt

### DISCUSSION THREAD

**Discussion by harbor6688** [mullvad/mullvadvpn-app#9006] [Feature requests & Ideas]
### I have checked if others have suggested this already

- [X] I have checked this issue tracker to see if others have reported similar issues.

### Feature description

I wonder if it would be possible to provide a cli only package for mullvad-vpn for the open source router os openwrt the current solution using native wireguard support while working does miss a lot of advanced features like multihop quantum resistant etc.
The cli app probably would only need to handle the location choosing and advanced features since firewall rules can be easily handled in the webinterface by the user trough vpn intefaces would also have to be permanent and not generated dynamically only on the fly to allow for permanent firewall rules that constrain traffic towards that inteface.

### Alternative solutions

I am aware of the current solution of using the native wireguard/openvpn implementation in openwrt but those do not allow for some of the more advanced features the vpn package provides.

### Type of feature

- [X] Better privacy/anonymity
- [ ] Better at circumventing censorship
- [X] Easier to use
- [ ] Other

### Operating System

- [ ] Android
- [ ] iOS
- [ ] Windows
- [ ] macOS
- [X] Linux

**Comment by ngude** [comment]
this is exactly what i came here to suggest, thank you, it would be incredibly useful to have these additional features available on the hardware.

**Comment by MarkusPettersson98** [comment]
Hi, and thanks for the suggestion! It is a very interesting idea, I will lift this internally :blush:

**Comment by cohesive-flight** [comment]
@MarkusPettersson98 Is your response to the other request also applicable here?

**Comment by MarkusPettersson98** [comment]
> @MarkusPettersson98 Is your response to the other request also applicable here?

If you refer to https://github.com/mullvad/mullvadvpn-app/issues/6828#issuecomment-2367966998, then no :blush:

**Comment by danielrod36** [comment]
I would really be interested in this. Have been trying to make an Ubuntu router with the express purpose of running Mullvad to circumvent local censorship. I know too that basically every expat living here would immediately switch from their current VPN provider to Mullvad if this is implemented, as the current solution to have network-wide VPN here is... imperfect at best.

**Comment by MarkusPettersson98** [comment]
Even though censorship sucks, I'm glad to hear that you would be excited about OpenWRT support :blush: I just want to let you know that we are also excited about this, but as usual we can not guarantee when this will be done. Stay tuned I guess :wink:

**Comment by christianmg99** [comment]
Hi @MarkusPettersson98 could you update on whether this is somewhere in the roadmap?? I'm working on a project to have an always-on vpn router and it would be so much better to benefit from Mullvad extra features instead of using standard wireguard.

**Comment by MarkusPettersson98** [comment]
Hi @christianmg99 👋 

We have not currently planned exactly when to add support for OpenWRT, but we are open to add support for it eventually! I don't dare to promise a launch date though.

Best regards 💛

**Comment by dkwo** [comment]
Just to clarify my own ignorance: what is currently missing, if one just builds from current repo, and skips electron related stuff?
It seems to me, at least on linux `aarch64`, one gets a working mullvad vpn daemon, and just needs to provide their own service to run it. Is such a setup missing some features like shadowsocks etc?

**Comment by syboxez** [comment]
Is there a way to compile only the CLI application for mipsel already, ignoring Electron entirely? I am also confused as to whether this is already possible or not.

**Comment by dkwo** [comment]
It is certainly possible to skip electron gui and build daemon only.
E.g. for runit, I write my own service, and interact with daemon via the
cli command.  Everything seems to work fine, been testing for a while
on Void Linux (musl should be supported as of 2025.8-beta1).

For the architectures: only aarch64 and x86_64 are officially supported.
Based on my experirence, most of the stuff is rust, so that boils down to
whether mipsl is supported by rust.  The exception is go-wireguard, where
they by hand invoke the compiler for the given arch.  That may require a
small change, but again it will boil down to whether go supports mipsl,
or whatever arch you are interested in.

**Comment by MahdiNazemi** [comment]
With Mullvad ending OpenVPN support next year, and considering that most circumvention methods on OpenWRT currently rely on OpenVPN, it would be valuable to add support for the Mullvad CLI, as well as include protocols like Shadowsocks, QUIC, and UDP-over-TCP, which are already available in the desktop apps.

**Reply by User00123** [reply]
Hi, I see an openwrt branch in Mullvad code.
https://github.com/mullvad/mullvadvpn-app/tree/openwrt/openwrt

**Comment by ghost** [comment]
@MarkusPettersson98 is it currently possible to simulate "shadowsocks for wireguard" on openwrt/android (without mullvad app)?

is it possible to use lwo? is it possible to use quic on openwrt?

**Comment by braydonf** [comment]
Related discussions: https://github.com/mullvad/mullvadvpn-app/discussions/9604 (CLI without the GUI packages) and https://github.com/mullvad/mullvadvpn-app/discussions/10015 (RISC-V Builds)

**Comment by syboxez** [comment]
Support for `mips` would be nice as well. Here is a `uname -a` and `cat /proc/cpuinfo` from my router:

```
# uname -a
Linux OpenWrt 6.6.119 #0 SMP Wed Dec 17 21:08:22 2025 mips GNU/Linux
```

```
# ldd --version
musl libc (mipsel-sf)
Version 1.2.5
Dynamic Program Loader
Usage: ldd [options] [--] pathname
```

```
# cat /proc/cpuinfo # Truncated to a single CPU core even though it has 4 CPU cores
system type		: MediaTek MT7621 ver:1 eco:3
machine			: D-Link DIR-2055 A1
processor		: 0
cpu model		: MIPS 1004Kc V2.15
BogoMIPS		: 586.13
wait instruction	: yes
microsecond timers	: yes
tlb_entries		: 32
extra interrupt vector	: yes
hardware watchpoint	: yes, count: 4, address/irw mask: [0x0ffc, 0x0ffc, 0x0ffb, 0x0ffb]
isa			: mips1 mips2 mips32r1 mips32r2
ASEs implemented	: mips16 dsp mt
Options implemented	: tlb 4kex 4k_cache prefetch mcheck ejtag llsc pindexed_dcache userlocal vint perf_cntr_intr_bit cdmm perf
shadow register sets	: 1
kscratch registers	: 0
package			: 0
core			: 0
VPE			: 0
VCED exceptions		: not available
VCEI exceptions		: not available
```

**Comment by mohsen2hasani** [comment]
I don't know if it will be helpful, but I developed (edit: with Claude AI) an LWO support package for OpenWrt:
[LWO Support for OpenWrt](https://github.com/mohsen2hasani/LWO)

I created this because I needed to use a VPN on my webOS TV, but the operating system doesn’t support VPNs. Since I have a router running OpenWrt, I spent a weekend developing this to solve the problem—with the help of AI, I should mention!

**Reply by cgytrus** [reply]
this looks ai generated to me.

**Reply by mohsen2hasani** [reply]
Yes, 100%!
And it works on my home router, so what's the problem? 😁

**Reply by cgytrus** [reply]
the problem is that you didnt "develop" anything. and i have a hard time believing you understand any of the code the llm has written considering this is your only rust repo, and mullvad has policy against that.

**Reply by mohsen2hasani** [reply]
Fair enough, I'll mention this on the repo and update my comment above as well 🙏 
But I have some experience with Rust from a project, and I reviewed the code.
